packages/components/src/popover/index.js

```diff
@@ -170,20 +170,36 @@ const Popover = (
 		}
 
 		const iframeRect = frameElement.getBoundingClientRect();
-		return {
-			name: 'iframeOffset',
-			fn( { x, y } ) {
-				return {
-					x: x + iframeRect.left,
-					y: y + iframeRect.top,
-				};
-			},
-		};
+		return { x: iframeRect.left, y: iframeRect.top };
 	}, [ ownerDocument ] );
 
 	const middlewares = [
-		frameOffset,
 		offset ? offsetMiddleware( offset ) : undefined,
+		frameOffset
+			? offsetMiddleware( ( { placement: currentPlacement } ) => {
+					const isTopBottomPlacement =
```

Comment:
I feel we're getting closer with this approach, thanks for iterating!

Do you think we can only use once the `offset` middleware making checks for both `iframe` and passed `offset`. Currently if we are in an iframe and an `offset` is passed is not taken into the calculations.

Reply:
> Currently if we are in an iframe and an offset is passed, it's not included in the calculations.

That's interesting. I thought that adding two separate instances of the `offset` middleware would get around exactly that problem.

Combining the offset prop into the `iframeOffset` middleware would not be trivial, since `offset` can be both a simple number or an object with the (optional) `mainAxis`, `crossAxis` and `alignmentAxis` properties

Reply:
That's the internal implementation though. I see that we don't have in the README the `offset` property. Should we limit our Popover's component offset to just a number? Do you know if we supported something different in the old implementation?

Reply:
Good catch Nik.

I like the suggestion of treating Popover's offset as just a number for now.

I've pushed a commit that documents `offset` and uses only a single offset middleware.

Reply:
> It looks like before the offset was achieved using padding around the popover. There wasn't a prop. Dropdown is one of the notable components that switched to using the `offset` prop instead of css: https://github.com/WordPress/gutenberg/pull/40740/files#r869586527
> 
> I think the offset css was implemented as part of the Popover's styles, so there might be some instances of popovers out there that are missing offsets.

Good point. We should look for these overrides and see if they are compatible with the signature of the `offset` prop (i.e. a `number`) — there may be instances in which we may have to apply the offset only on one axis, for example.

Reply:
These are the styles for the offsets when popovers have no arrows:
https://github.com/WordPress/gutenberg/blob/b49948774e6d1e451d75c9df2e5763f95cb7e965/packages/components/src/popover/style.scss#L148-L163

A `number` property should suffice in replacing those.

When the popover has an arrow, there are slightly different offsets used, including a slight lateral shift:
https://github.com/WordPress/gutenberg/blob/b49948774e6d1e451d75c9df2e5763f95cb7e965/packages/components/src/popover/style.scss#L29

And many more complicated styles below that line. I'm not sure why this CSS exists, possibly to centre the arrow? I don't think this is necessary now, floating UI has its own arrow middleware. In `trunk` arrows only seem to work for bottom placements—I think it'd be a good idea to treat arrow improvements as a separate bug. Gutenberg doesn't have any popovers with arrows, and it's purely a cosmetic problem, so it's a much less urgent issue to fix IMO.

Reply:
> These are the styles for the offsets when popovers have no arrows:

It looks like you're pointing to an outdated version of the code — I can't see those margin styles in the popover styles anymore.

But I otherwise agree with the points that you make!

> A `number` property should suffice in replacing those.

Agreed! We should then replace all instances across the repository where we're applying a `margin` in CSS, and replace it with the `offset` prop.

Arrow placement is currently achieved via the [dedicated floating ui middleware](https://github.com/WordPress/gutenberg/blob/2947844bde8f8e077677c4f2a695184325244198/packages/components/src/popover/index.js#L225), plus [a few custom styles](https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/components/src/popover/style.scss#L61-L71) (which mostly following [the suggested approach in floating UI docs](https://floating-ui.com/docs/tutorial#arrow-middleware)). But we can definitely look into refining it (if needed) in a follow-up PR.

Reply:
> It looks like you're pointing to an outdated version of the code — I can't see those margin styles in the popover styles anymore.

Sorry if that wasn't clear. The code is from the commit before Riad's refactor. The margins were all removed in that PR and some were changed to use the offset prop (like `Dropdown`), but many popovers now seem to have no gap between floating and reference elements.

Reply:
> The code is from the commit before Riad's refactor [...] but many popovers now seem to have no gap between floating and reference elements.

Gotcha! Yeah, it sounds like we'll need to use the `offset` prop in those scenarios.

But also, I quickly searched for `components-popover` to find all style overrides in the repo, and there's definitely a few cases where `margin` is being used ([example](https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/block-editor/src/components/block-switcher/style.scss#L104), [example](https://github.com/WordPress/gutenberg/blob/ff3eaab5f991b7b84ee46ac5449794342f4772aa/packages/block-editor/src/components/block-tools/style.scss#L194-L195)) and should likely be replaced with `offset`

Reply:
I couldn't see any that seem like they'd be a problem, but there might be some that are using classnames properly 😅 

This PR also doesn't seem to cause any issues with using classnames, so I think it's ok to ship and migrate offsets over to the prop in follow-up PR.

I also want to switch a lot of the codebase to use `placement` instead of `position` as well. There's no point running everything through that conversion function at runtime.

Reply:
Good point. I've decided to open an issue to track all of these enhancements that we'd like to apply to the component — feel free to edit the issue's original description or to add a comment there!